cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogFileManager.java

```diff
@@ -0,0 +1,207 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class including logic for getting log file to write to. Used by {@link CDAPLogAppender}
+ */
+class LogFileManager implements Flushable {
+  private static final Logger LOG = LoggerFactory.getLogger(LogFileManager.class);
+
+  private final long maxLifetimeMillis;
+  private final Map<LogPathIdentifier, LogFileOutputStream> outputStreamMap;
+  private final Location logsDirectoryLocation;
+  private final FileMetaDataManager fileMetaDataManager;
+  private final int syncIntervalBytes;
+  private final Schema schema;
+
+  LogFileManager(long maxFileLifetimeMs, int syncIntervalBytes, Schema schema,
+                 FileMetaDataManager fileMetaDataManager,
+                 LocationFactory locationFactory) {
+    this.maxLifetimeMillis = maxFileLifetimeMs;
+    this.syncIntervalBytes = syncIntervalBytes;
+    this.schema = schema;
+    this.fileMetaDataManager = fileMetaDataManager;
+    this.logsDirectoryLocation = locationFactory.create("logs");
+    this.outputStreamMap = new HashMap<>();
+  }
+
+  /**
+   * Get log file output stream for the give logging context and timestamp - return an already open file,
+   * or create and return a new file or rotate a file based on time and return the corresponding stream.
+   * @param logPathIdentifier identify logging context
+   * @param eventTimestamp timestamp from logging event
+   * @return LogFileOutputStream output stream to the log file
+   * @throws IOException if there is exception while getting location or while writing meta data
+   */
+  public LogFileOutputStream getLogFileOutputStream(LogPathIdentifier logPathIdentifier,
+                                                    long eventTimestamp) throws IOException {
+    LogFileOutputStream logFileOutputStream = outputStreamMap.get(logPathIdentifier);
+    if (logFileOutputStream == null) {
+      logFileOutputStream = createOutputStream(logPathIdentifier, eventTimestamp);
+    } else {
+      // rotate the file if needed (time has passed)
+      logFileOutputStream = rotateOutputStream(logFileOutputStream, logPathIdentifier, eventTimestamp);
+    }
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream createOutputStream(final LogPathIdentifier identifier,
+                                                 long timestamp) throws IOException {
+    TimeStampLocation location = createLocation(identifier);
+    try {
+      fileMetaDataManager.writeMetaData(identifier, timestamp, location.getTimeStamp(), location.getLocation());
+    } catch (Throwable e) {
+      // delete created file as there was exception while writing meta data
+      Locations.deleteQuietly(location.getLocation());
+      throw new IOException(e);
+    }
+
+    LOG.info("Created Avro file at {}", location);
+    LogFileOutputStream logFileOutputStream = new LogFileOutputStream(
+      location.getLocation(), schema, syncIntervalBytes, new Closeable() {
```

Comment:
Should have `location.getTimeStamp()` pass to `LogFileOutputStream` as well instead of calling `System.currentTimeMillis()` in the constructor.